File: providers/security/packs/file-boundaries.mjs
// File-boundary security pack: path traversal, archive traversal (zip-slip),
// and unchecked symlink following. Every rule is a lexical (pattern-only)
// detector: it never opens a file, never extracts an archive, never calls a
// model, and never certifies a finding. It only ever emits UNADJUDICATED
// candidates via the sealed candidate-engine. When a normalization /
// containment-check signal is observed on the same path, the candidate is
// downgraded with the observed control referenced in `observedControls` (a
// model control entity was found) or noted in `uncertainty` (only a lexical
// mitigating signal was found).

import { digest } from '../contracts.mjs';

function findArtifact(context, file) {
  return context.model.entities.find((e) => e.kind === 'repository-artifact' && e.attributes?.path === file);
}

function findRelatedControl(context, artifactId, controlTypes) {
  if (artifactId) {
    for (const rel of context.relationsTo(artifactId)) {
      if (rel.kind !== 'protects') continue;
      const control = context.entityById.get(rel.from);
      if (control?.kind === 'control' && controlTypes.includes(control.attributes?.controlType)) return control;
    }
  }
  return context.model.entities.find((e) => e.kind === 'control' && controlTypes.includes(e.attributes?.controlType)) ?? null;
}

function lineOf(text, index) {
  return text.slice(0, index).split('\n').length;
}

function windowAround(text, index, matchLength, radius) {
  const start = Math.max(0, index - radius);
  const end = Math.min(text.length, index + matchLength + radius);
  return text.slice(start, end);
}

function testWindow(pattern, text, index, matchLength, radius) {
  if (!pattern) return false;
  return pattern.test(windowAround(text, index, matchLength, radius));
}

const CANDIDATE_CLASS = 'file-boundaries';

const RULES = [
  {
    id: 'file.path-traversal-unvalidated',
    pattern: /\b(fs\.readFile(?:Sync)?|fs\.writeFile(?:Sync)?|fs\.createReadStream|fs\.createWriteStream|fs\.open(?:Sync)?|res\.sendFile|res\.download)\s*\(\s*([^()\n]*\b(?:request|req)\.(?:query|body|params)(?:\.[\w]+)?\b[^()\n]*)\)/gi,
    sinkApiOf: (m) => m[1],
    sourceExprOf: (m) => m[2].trim(),
    claim: 'Request-controlled path data reaches a filesystem operation without a visible normalization/containment check, risking path traversal outside the intended directory.',
    severityHint: 'high',
    attackerCapabilities: ['control-request-input'],
    preconditionAction: 'supply-path',
    environment: 'application',
    effectKind: 'data-access', effectAction: 'access', effectScope: 'filesystem-outside-intended-directory',
    chainRoles: ['starter', 'impact'],
    downgrade: {
      controlTypes: ['path-traversal-containment', 'path-normalization'],
      severityHint: 'low',
      lexicalPattern: /path\.normalize|path\.resolve|\.startsWith\(|sanitize-filename|sanitizeFilename|path\.basename/i,
      lexicalNote: 'A path normalization/containment check is present near the sink; treated as a mitigating signal pending adjudication.',
      radius: 300,
    },
    uncertainty: ['A request-controlled path argument is not automatically traversal; whether the value is normalized and confined to an intended base directory elsewhere in the call path must be adjudicated.'],
  },
  {
    id: 'file.archive-zip-slip',
    pattern: /\b(fs\.writeFile(?:Sync)?|fs\.createWriteStream|path\.join)\s*\(\s*([^()\n]*\b(?:entry|zipEntry|file)\.(?:path|fileName|name)\b[^()\n]*)\)/gi,
    sinkApiOf: (m) => m[1],
    sourceExprOf: (m) => m[2].trim(),
    claim: 'An archive entry path is written to the filesystem without a visible containment check against the extraction directory, risking a zip-slip path-traversal write.',
    severityHint: 'high',
    attackerCapabilities: ['supply-malicious-archive'],
    preconditionAction: 'supply-archive-entry',
    environment: 'application',
    effectKind: 'integrity-impact', effectAction: 'overwrite', effectScope: 'filesystem-outside-extraction-directory',
    chainRoles: ['starter', 'impact'],
    downgrade: {
      controlTypes: ['zip-slip-containment', 'archive-path-validation'],
      severityHint: 'low',
      lexicalPattern: /path\.normalize|\.startsWith\(|resolvedPath|isInsideDir|zip-slip|sanitizeEntry/i,
      lexicalNote: 'A resolved-path containment check is present near the sink; treated as a mitigating signal pending adjudication.',
      radius: 300,
    },
    uncertainty: ['Whether the archive source is attacker-controlled, and whether the resolved entry path is validated to remain inside the extraction directory, must be adjudicated.'],
  },
  {
    id: 'file.symlink-unchecked',
    pattern: /\b(followSymlinks\s*:\s*true|dereference\s*:\s*true)\b/g,
    sinkApiOf: () => 'archive-extraction-symlink-option',
    sourceExprOf: (m) => m[1],
    claim: 'Archive/file extraction is configured to follow symbolic links, allowing an entry to write through a symlink to a location outside the intended target directory.',
    severityHint: 'medium',
    attackerCapabilities: ['supply-malicious-archive'],
    preconditionAction: 'supply-symlink-entry',
    environment: 'application',
    effectKind: 'integrity-impact', effectAction: 'traverse-symlink', effectScope: 'filesystem-outside-intended-directory',
    chainRoles: ['starter', 'impact'],
    downgrade: {
      controlTypes: ['symlink-containment-check'],
      severityHint: 'low',
      lexicalPattern: /isSymbolicLink\(\)|lstat|realpath[\s\S]{0,80}startsWith/i,
      lexicalNote: 'A symlink/realpath containment check is present near the extraction option; treated as a mitigating signal pending adjudication.',
      radius: 300,
    },
    uncertainty: ['Whether the archive/file source can be attacker-controlled, and whether a symlink-containment check exists elsewhere in the extraction pipeline, must be adjudicated.'],
  },
];

function analyze(context) {
  const observations = [];
  for (const file of context.files) {
    const text = context.readFile(file) ?? '';
    if (!text) continue;
    const artifact = findArtifact(context, file);
    for (const rule of RULES) {
      rule.pattern.lastIndex = 0;
      let match;
      while ((match = rule.pattern.exec(text)) !== null) {
        if (match[0].length === 0) { rule.pattern.lastIndex += 1; continue; }

        let severityHint = rule.severityHint;
        let observedControls = [];
        let controlObserved = null;
        const uncertainty = [...rule.uncertainty];

        const control = findRelatedControl(context, artifact?.id, rule.downgrade.controlTypes);
        if (control) {
          severityHint = rule.downgrade.severityHint;
          observedControls = [control.id];
          controlObserved = control.name;
          uncertainty.push(`Observed ${control.attributes?.controlType ?? 'mitigating'} control (${control.name}) on this path; downgraded pending adjudication of coverage completeness.`);
        } else if (testWindow(rule.downgrade.lexicalPattern, text, match.index, match[0].length, rule.downgrade.radius)) {
          severityHint = rule.downgrade.severityHint;
          controlObserved = 'lexical-signal';
          uncertainty.push(rule.downgrade.lexicalNote);
        }

        observations.push({
          ruleId: rule.id,
          candidateClass: CANDIDATE_CLASS,
          claim: rule.claim,
          severityHint,
          sources: artifact ? [artifact.id] : [],
          sinks: artifact ? [artifact.id] : [],
          attackerCapabilities: rule.attackerCapabilities,
          preconditions: [{
            kind: 'attacker-position', subject: 'actor:external', action: rule.preconditionAction,
            scope: null, environment: rule.environment, tenant: null,
          }],
          effects: [{
            kind: rule.effectKind, subject: 'actor:external', action: rule.effectAction,
            object: artifact?.id ?? null, scope: rule.effectScope, environment: rule.environment, tenant: null,
          }],
          assets: [], trustBoundaryCrossings: [], requiredControls: [], observedControls,
          chainRoles: rule.chainRoles,
          evidenceRefs: artifact?.evidenceRefs ?? [],
          detectorMetadata: {
            file,
            line: lineOf(text, match.index),
            sourceExpr: rule.sourceExprOf(match),
            sinkApi: rule.sinkApiOf(match),
            controlObserved,
            matchDigest: digest(match[0]),
          },
          uncertainty,
        });
      }
    }
  }
  return observations;
}

function buildVariantStrategy(rule) {
  return {
    rootCause(candidate) {
      return {
        class: `${rule.id}-unmitigated`,
        semanticFeatures: ['tainted-path-to-sink', rule.id, candidate.detectorMetadata?.sinkApi ?? rule.id],
      };
    },
    enumerate(context) {
      const matches = [];
      for (const file of context.files) {
        const text = context.readFile(file) ?? '';
        if (!text) continue;
        rule.pattern.lastIndex = 0;
        let match;
        while ((match = rule.pattern.exec(text)) !== null) {
          if (match[0].length === 0) { rule.pattern.lastIndex += 1; continue; }
          matches.push({
            file,
            line: lineOf(text, match.index),
            semanticFingerprint: digest({ ruleId: rule.id, file, snippet: match[0] }),
            disposition: 'CONFIRMED',
          });
        }
      }
      return {
        denominator: { kind: 'source-files', digest: context.denominatorDigest, expected: context.files.length, examined: context.files.length, unexamined: [] },
        strategies: [{
          id: `${rule.id}-search`, kind: 'lexical-fallback',
          description: `Enumerate every ${rule.id} occurrence across the denominator.`,
          queryDigest: digest(rule.id), complete: true, coverageGaps: [],
        }],
        matches,
        coverageGaps: [],
      };
    },
  };
}

export default Object.freeze({
  id: 'security.file-boundaries',
  version: '1.0.0',
  candidateClass: CANDIDATE_CLASS,
  description: 'Path traversal, archive traversal (zip-slip), and unchecked symlink following.',
  rules: RULES.map((rule) => ({ id: rule.id })),
  analyze,
  variantStrategies: Object.fromEntries(RULES.map((rule) => [rule.id, buildVariantStrategy(rule)])),
});

File: engine/crates/legion-audit/src/wf_port/wf059/file_boundaries.rs
//! Port of `src/providers/security/packs/file-boundaries.mjs`.
//!
//! Path traversal, archive traversal (zip-slip), and unchecked symlink
//! following. When a normalization/containment-check signal is observed on
//! the same path, the candidate is downgraded with the observed control
//! referenced in `observedControls` (a model control entity was found) or
//! noted in `uncertainty` (only a lexical mitigating signal was found).

use super::common::{digest, line_of, window_around, Context, Fact, Observation};
use regex::Regex;
use serde_json::json;
use std::sync::LazyLock;

static PATH_TRAVERSAL_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)\b(fs\.readFile(?:Sync)?|fs\.writeFile(?:Sync)?|fs\.createReadStream|fs\.createWriteStream|fs\.open(?:Sync)?|res\.sendFile|res\.download)\s*\(\s*([^()\n]*\b(?:request|req)\.(?:query|body|params)(?:\.[\w]+)?\b[^()\n]*)\)").unwrap()
});
static ZIP_SLIP_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)\b(fs\.writeFile(?:Sync)?|fs\.createWriteStream|path\.join)\s*\(\s*([^()\n]*\b(?:entry|zipEntry|file)\.(?:path|fileName|name)\b[^()\n]*)\)").unwrap()
});
static SYMLINK_UNCHECKED_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\b(followSymlinks\s*:\s*true|dereference\s*:\s*true)\b").unwrap());

static PATH_TRAVERSAL_DOWNGRADE_LEXICAL: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)path\.normalize|path\.resolve|\.startsWith\(|sanitize-filename|sanitizeFilename|path\.basename").unwrap());
static ZIP_SLIP_DOWNGRADE_LEXICAL: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)path\.normalize|\.startsWith\(|resolvedPath|isInsideDir|zip-slip|sanitizeEntry").unwrap());
static SYMLINK_DOWNGRADE_LEXICAL: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?is)isSymbolicLink\(\)|lstat|realpath.{0,80}startsWith").unwrap());

struct Rule {
    id: &'static str,
    claim: &'static str,
    severity_hint: &'static str,
    attacker_capabilities: &'static [&'static str],
    precondition_action: &'static str,
    effect_kind: &'static str,
    effect_action: &'static str,
    effect_scope: &'static str,
    chain_roles: &'static [&'static str],
    downgrade_control_types: &'static [&'static str],
    downgrade_severity_hint: &'static str,
    downgrade_lexical: &'static LazyLock<Regex>,
    downgrade_radius: usize,
    downgrade_note: &'static str,
    uncertainty: &'static str,
}

const RULE_PATH_TRAVERSAL: Rule = Rule {
    id: "file.path-traversal-unvalidated",
    claim: "Request-controlled path data reaches a filesystem operation without a visible normalization/containment check, risking path traversal outside the intended directory.",
    severity_hint: "high",
    attacker_capabilities: &["control-request-input"],
    precondition_action: "supply-path",
    effect_kind: "data-access",
    effect_action: "access",
    effect_scope: "filesystem-outside-intended-directory",
    chain_roles: &["starter", "impact"],
    downgrade_control_types: &["path-traversal-containment", "path-normalization"],
    downgrade_severity_hint: "low",
    downgrade_lexical: &PATH_TRAVERSAL_DOWNGRADE_LEXICAL,
    downgrade_radius: 300,
    downgrade_note: "A path normalization/containment check is present near the sink; treated as a mitigating signal pending adjudication.",
    uncertainty: "A request-controlled path argument is not automatically traversal; whether the value is normalized and confined to an intended base directory elsewhere in the call path must be adjudicated.",
};

const RULE_ZIP_SLIP: Rule = Rule {
    id: "file.archive-zip-slip",
    claim: "An archive entry path is written to the filesystem without a visible containment check against the extraction directory, risking a zip-slip path-traversal write.",
    severity_hint: "high",
    attacker_capabilities: &["supply-malicious-archive"],
    precondition_action: "supply-archive-entry",
    effect_kind: "integrity-impact",
    effect_action: "overwrite",
    effect_scope: "filesystem-outside-extraction-directory",
    chain_roles: &["starter", "impact"],
    downgrade_control_types: &["zip-slip-containment", "archive-path-validation"],
    downgrade_severity_hint: "low",
    downgrade_lexical: &ZIP_SLIP_DOWNGRADE_LEXICAL,
    downgrade_radius: 300,
    downgrade_note: "A resolved-path containment check is present near the sink; treated as a mitigating signal pending adjudication.",
    uncertainty: "Whether the archive source is attacker-controlled, and whether the resolved entry path is validated to remain inside the extraction directory, must be adjudicated.",
};

const RULE_SYMLINK: Rule = Rule {
    id: "file.symlink-unchecked",
    claim: "Archive/file extraction is configured to follow symbolic links, allowing an entry to write through a symlink to a location outside the intended target directory.",
    severity_hint: "medium",
    attacker_capabilities: &["supply-malicious-archive"],
    precondition_action: "supply-symlink-entry",
    effect_kind: "integrity-impact",
    effect_action: "traverse-symlink",
    effect_scope: "filesystem-outside-intended-directory",
    chain_roles: &["starter", "impact"],
    downgrade_control_types: &["symlink-containment-check"],
    downgrade_severity_hint: "low",
    downgrade_lexical: &SYMLINK_DOWNGRADE_LEXICAL,
    downgrade_radius: 300,
    downgrade_note: "A symlink/realpath containment check is present near the extraction option; treated as a mitigating signal pending adjudication.",
    uncertainty: "Whether the archive/file source can be attacker-controlled, and whether a symlink-containment check exists elsewhere in the extraction pipeline, must be adjudicated.",
};

fn matches_for(rule: &Rule, ctx: &Context) -> Vec<(String, usize, usize, usize, String, String, String)> {
    let mut out = Vec::new();
    for file in &ctx.files {
        let text = ctx.read_file(file);
        if text.is_empty() {
            continue;
        }
        let (source_of, sink_of, captures): (
            fn(&regex::Captures<'_>) -> String,
            fn(&regex::Captures<'_>) -> String,
            Vec<regex::Captures<'_>>,
        ) =
            match rule.id {
                "file.path-traversal-unvalidated" => (
                    |c| c.get(2).map(|g| g.as_str().trim().to_string()).unwrap_or_default(),
                    |c| c.get(1).map(|g| g.as_str().to_string()).unwrap_or_default(),
                    PATH_TRAVERSAL_PATTERN.captures_iter(text).collect(),
                ),
                "file.archive-zip-slip" => (
                    |c| c.get(2).map(|g| g.as_str().trim().to_string()).unwrap_or_default(),
                    |c| c.get(1).map(|g| g.as_str().to_string()).unwrap_or_default(),
                    ZIP_SLIP_PATTERN.captures_iter(text).collect(),
                ),
                "file.symlink-unchecked" => (
                    |c| c.get(1).map(|g| g.as_str().to_string()).unwrap_or_default(),
                    |_| "archive-extraction-symlink-option".to_string(),
                    SYMLINK_UNCHECKED_PATTERN.captures_iter(text).collect(),
                ),
                _ => (|_| String::new(), |_| String::new(), Vec::new()),
            };
        for m in captures {
            let whole = m.get(0).unwrap();
            out.push((
                file.clone(),
                whole.start(),
                whole.len(),
                line_of(text, whole.start()),
                source_of(&m),
                sink_of(&m),
                whole.as_str().to_string(),
            ));
        }
    }
    out
}

/// Faithful port of `analyze(context)`.
pub fn analyze(ctx: &Context) -> Vec<Observation> {
    let mut out = Vec::new();
    for rule in [&RULE_PATH_TRAVERSAL, &RULE_ZIP_SLIP, &RULE_SYMLINK] {
        for (file, index, len, line, source_expr, sink_api, matched) in matches_for(rule, ctx) {
            let text = ctx.read_file(&file);
            let artifact = ctx.find_artifact(&file);

            let mut severity_hint = rule.severity_hint.to_string();
            let mut observed_controls: Vec<String> = vec![];
            let mut control_observed: Option<String> = None;
            let mut uncertainty = vec![rule.uncertainty.to_string()];

            let control = artifact.and_then(|a| {
                ctx.relations_to(&a.id)
                    .filter(|r| r.kind == "protects")
                    .find_map(|r| {
                        ctx.entity_by_id(&r.from).filter(|c| {
                            c.kind == "control"
                                && c.attr_str("controlType")
                                    .map(|t| rule.downgrade_control_types.contains(&t))
                                    .unwrap_or(false)
                        })
                    })
            }).or_else(|| {
                ctx.entities.iter().find(|e| {
                    e.kind == "control"
                        && e.attr_str("controlType")
                            .map(|t| rule.downgrade_control_types.contains(&t))
                            .unwrap_or(false)
                })
            });

            if let Some(control) = control {
                severity_hint = rule.downgrade_severity_hint.to_string();
                observed_controls = vec![control.id.clone()];
                let control_type = control.attr_str("controlType").unwrap_or("mitigating");
                let control_name = control.attr_str("name").unwrap_or(&control.id);
                control_observed = Some(control_name.to_string());
                uncertainty.push(format!(
                    "Observed {control_type} control ({control_name}) on this path; downgraded pending adjudication of coverage completeness."
                ));
            } else if rule.downgrade_lexical.is_match(window_around(text, index, len, rule.downgrade_radius)) {
                severity_hint = rule.downgrade_severity_hint.to_string();
                control_observed = Some("lexical-signal".to_string());
                uncertainty.push(rule.downgrade_note.to_string());
            }

            out.push(Observation {
                rule_id: rule.id.to_string(),
                candidate_class: "file-boundaries".to_string(),
                claim: rule.claim.to_string(),
                severity_hint,
                sources: artifact.map(|a| vec![a.id.clone()]).unwrap_or_default(),
                sinks: artifact.map(|a| vec![a.id.clone()]).unwrap_or_default(),
                attacker_capabilities: rule.attacker_capabilities.iter().map(|s| s.to_string()).collect(),
                preconditions: vec![Fact {
                    kind: "attacker-position".to_string(),
                    subject: "actor:external".to_string(),
                    action: rule.precondition_action.to_string(),
                    object: None,
                    scope: None,
                    environment: "application".to_string(),
                    tenant: None,
                }],
                effects: vec![Fact {
                    kind: rule.effect_kind.to_string(),
                    subject: "actor:external".to_string(),
                    action: rule.effect_action.to_string(),
                    object: artifact.map(|a| a.id.clone()),
                    scope: Some(rule.effect_scope.to_string()),
                    environment: "application".to_string(),
                    tenant: None,
                }],
                assets: vec![],
                trust_boundary_crossings: vec![],
                required_controls: vec![],
                observed_controls,
                chain_roles: rule.chain_roles.iter().map(|s| s.to_string()).collect(),
                evidence_refs: Vec::new(),
                detector_metadata: json!({
                    "file": file,
                    "line": line,
                    "sourceExpr": source_expr,
                    "sinkApi": sink_api,
                    "controlObserved": control_observed,
                    "matchDigest": digest(&matched),
                }),
                uncertainty,
            });
        }
    }
    out
}

pub const PACK_ID: &str = "security.file-boundaries";
pub const PACK_VERSION: &str = "1.0.0";
pub const CANDIDATE_CLASS: &str = "file-boundaries";
